src/task.c

```diff
@@ -747,25 +686,116 @@ task_group_perform_one(task_group *group)
 platform_status
 task_perform_one(task_system *ts)
 {
-   platform_assert(!ts->use_bg_threads);
    platform_status rc = STATUS_OK;
    for (task_type type = TASK_TYPE_FIRST; type != NUM_TASK_TYPES; type++) {
-      rc = task_group_perform_one(&ts->group[type]);
+      rc = task_group_perform_one(&ts->group[type], 0);
+      /* STATUS_TIMEDOUT means no task was waiting. */
+      if (STATUS_IS_NE(rc, STATUS_TIMEDOUT)) {
+         return rc;
+      }
+   }
+   return rc;
+}
+
+platform_status
+task_perform_one_if_needed(task_system *ts)
+{
+   platform_status rc = STATUS_OK;
+   for (task_type type = TASK_TYPE_FIRST; type != NUM_TASK_TYPES; type++) {
+      /* Perform a task only if there are more waiting tasks than bg
+         threads. */
+      rc = task_group_perform_one(&ts->group[type],
+                                  ts->group[type].bg.num_threads);
+      /* STATUS_TIMEDOUT means no task was waiting. */
       if (STATUS_IS_NE(rc, STATUS_TIMEDOUT)) {
          return rc;
       }
    }
    return rc;
 }
 
+void
+task_perform_all(task_system *ts)
+{
+   platform_status rc;
+   do {
+      rc = task_perform_one(ts);
+      debug_assert(SUCCESS(rc) || STATUS_IS_EQ(rc, STATUS_TIMEDOUT));
+      /* STATUS_TIMEDOUT means no task was waiting. */
+   } while (STATUS_IS_NE(rc, STATUS_TIMEDOUT));
+}
+
+bool
+task_system_is_quiescent(task_system *ts)
+{
+   platform_status rc;
+   task_type       ttlocked;
+   bool            result = FALSE;
+
+   for (ttlocked = TASK_TYPE_FIRST; ttlocked < NUM_TASK_TYPES; ttlocked++) {
+      rc = platform_condvar_lock(&ts->group[ttlocked].cv);
+      if (!SUCCESS(rc)) {
+         goto cleanup;
+      }
+   }
+
+   for (task_type ttchecked = TASK_TYPE_FIRST; ttchecked < NUM_TASK_TYPES;
+        ttchecked++)
+   {
+      if (ts->group[ttchecked].current_executing_tasks
+          || ts->group[ttchecked].current_waiting_tasks)
+      {
+         goto cleanup;
+      }
+   }
+   result = TRUE;
+
+cleanup:
+   for (task_type tt = TASK_TYPE_FIRST; tt < ttlocked; tt++) {
+      rc = platform_condvar_unlock(&ts->group[tt].cv);
+      debug_assert(SUCCESS(rc));
+   }
+
+   return result;
+}
+
+void
+task_perform_until_quiescent(task_system *ts)
+{
+   while (!task_system_is_quiescent(ts)) {
+      task_perform_all(ts);
+   }
+}
+
+platform_status
+task_system_config_init(task_system_config *task_cfg,
+                        bool                use_stats,
+                        const uint64        num_bg_threads[NUM_TASK_TYPES],
+                        uint64              scratch_size)
+{
+   if ((num_bg_threads[TASK_TYPE_NORMAL] == 0)
+       != (num_bg_threads[TASK_TYPE_MEMTABLE] == 0))
+   {
+      return STATUS_BAD_PARAM;
```

Comment:
This refactoring to check `task_system_config_init()` separately here loses the error message that the other fn, `task_config_valid()` calls on L 801.

I think it will be helpful to move that error message here so that we don't get a silent failure due to an illegal configuration.

Reply:
Also, `task_config_valid()` checks for:

```
 810    if ((normal_bg_threads + memtable_bg_threads) >= MAX_THREADS) {
 811       platform_error_log("Total number of background threads configured"
```

Don't you want to move that check here? 

If the expectation is that anywhere in Splinter main code or in the test code, before we call `task_system_create()` that the code would [should] have called your new function `task_system_config_init()`, then we could do something like this:

- Have `task_system_config_init()` call `task_config_valid()` to do all the required checks. The callee will do the required error reporting.
- then once validation succeeds, continue with `task_system_config_init()`.